lib/zoneinfo_module.c

```diff
@@ -0,0 +1,2194 @@
+             long dst_offset, _TransitionRuleType *start,
+             _TransitionRuleType *end, _tzrule *out);
+static void
+free_tzrule(_tzrule *tzrule);
+
+static PyObject *
+load_timedelta(long seconds);
+
+static int
+get_local_timestamp(PyObject *dt, int64_t *local_ts);
+static _ttinfo *
+find_ttinfo(PyZoneInfo_ZoneInfo *self, PyObject *dt);
+
+static int
+ymd_to_ord(int y, int m, int d);
+static int
+is_leap_year(int year);
+
+static int
+_bisect(const int64_t value, const int64_t *arr, size_t size);
+
+static PyObject *
+zoneinfo_new_instance(PyTypeObject *type, PyObject *key)
+{
+    PyObject *file_obj = NULL;
+    PyObject *file_path = NULL;
+
+    file_path = PyObject_CallFunctionObjArgs(_tzpath_find_tzfile, key, NULL);
+    if (file_path == NULL) {
+        return NULL;
+    }
+    else if (file_path == Py_None) {
+        file_obj = PyObject_CallMethod(_common_mod, "load_tzdata", "O", key);
+        if (file_obj == NULL) {
+            return NULL;
+        }
+    }
+
+    PyObject *self = (PyObject *)(type->tp_alloc(type, 0));
+    if (self == NULL) {
+        goto error;
+    }
+
+    if (file_obj == NULL) {
+        file_obj = PyObject_CallFunction(io_open, "Os", file_path, "rb");
+        if (file_obj == NULL) {
+            goto error;
+        }
+    }
+
+    if (load_data((PyZoneInfo_ZoneInfo *)self, file_obj)) {
+        goto error;
+    }
+
+    PyObject *rv = PyObject_CallMethod(file_obj, "close", NULL);
+    Py_DECREF(file_obj);
+    file_obj = NULL;
+    if (rv == NULL) {
+        goto error;
+    }
+    Py_DECREF(rv);
+
+    ((PyZoneInfo_ZoneInfo *)self)->key = key;
+    Py_INCREF(key);
+
+    goto cleanup;
+error:
+    Py_XDECREF(self);
+    self = NULL;
+cleanup:
+    if (file_obj != NULL) {
+        PyObject_CallMethod(file_obj, "close", NULL);
+        Py_DECREF(file_obj);
+    }
+    Py_DECREF(file_path);
+    return self;
+}
+
+static PyObject *
+zoneinfo_new(PyTypeObject *type, PyObject *args, PyObject *kw)
+{
+    // TODO: Support subclasses
+    PyObject *key = NULL;
+    static char *kwlist[] = {"key", NULL};
+    if (PyArg_ParseTupleAndKeywords(args, kw, "O", kwlist, &key) == 0) {
+        return NULL;
+    }
+
+    assert(ZONEINFO_WEAK_CACHE != NULL);
+    PyObject *instance =
+        PyObject_CallMethod(ZONEINFO_WEAK_CACHE, "get", "O", key, Py_None);
+    if (instance == NULL) {
+        return NULL;
+    }
+
+    if (instance == Py_None) {
+        PyObject *setdefault = PyUnicode_FromString("setdefault");
+        if (setdefault == NULL) {
+            return NULL;
+        }
+
+        PyObject *tmp = zoneinfo_new_instance(type, key);
+        if (tmp == NULL) {
+            return NULL;
+        }
+
+        instance = PyObject_CallMethodObjArgs(ZONEINFO_WEAK_CACHE, setdefault,
+                                              key, tmp, NULL);
+        ((PyZoneInfo_ZoneInfo *)instance)->source = SOURCE_CACHE;
+
+        Py_DECREF(tmp);
+        Py_DECREF(setdefault);
```

Comment:
Shouldn't there be `Py_DECREF(setdefault)` if the `if (tmp == NULL)` block? E.g.:

```c
        PyObject *tmp = zoneinfo_new_instance(type, key);
        if (tmp == NULL) {
            Py_DECREF(setdefault);
            return NULL;
        }
```

Otherwise, the above seems like a possible refleak in the rare situation that `tmp` is NULL, but `setdefault` isn't. Such as if the object `setdefault` is pointing to can be allocated, but not the object that `tmp` is pointing to.

(Note that my area of experience isn't the C-API, so the above may not be an concern in reality. It just caught my attention while I was looking over the code.)

Reply:
Good catch, thanks!

I switched this over to use `PyObject_CallMethod` instead of `PyObject_CallMethodObjArgs`, which makes it so I don't have to create a `setdefault` object. This code is only executed during a cache miss, in which case it's calling into pure Python code anyway and the difference between `PyObject_CallMethod` and `PyObject_CallMethodObjArgs` is negligible.